dali/pipeline/util/thread_pool.cc

```diff
@@ -56,26 +56,30 @@ ThreadPool::~ThreadPool() {
 }
 
 void ThreadPool::AddWork(Work work, int64_t priority, bool start_immediately) {
-  bool started_before = false;
-  {
+  bool started_before = started_;
+  if (started_before) {
     std::lock_guard<std::mutex> lock(mutex_);
     work_queue_.push({priority, std::move(work)});
-    outstanding_work_.fetch_add(1);
-    started_before = started_;
-    started_ |= start_immediately;
+  } else {
+    work_queue_.push({priority, std::move(work)});
+    if (start_immediately) {
+      std::lock_guard<std::mutex> lock(mutex_);
+      started_ = true;
+    }
   }
+  outstanding_work_.fetch_add(1);
```

Comment:
Actually, this should go before you add work to the queue. Otherwise we risk signaling completion prematurely.